File: comentados/relatorios.py
from flask import Blueprint, request, jsonify, make_response
from models.user import db
from models.cadastro.base import Cidade, Equipe, Material, RegistroMaterial
from datetime import datetime
import csv
import io
from fpdf2 import FPDF

relatorio_bp = Blueprint('relatorio', __name__)

@relatorio_bp.route('/material-por-dia', methods=['GET'])
def relatorio_material_por_dia():
    """Gera relatório de uso de materiais agrupados por dia e equipe"""
    # Parâmetros de filtro
    formato = request.args.get('formato', 'json')  # json, csv ou pdf
    equipe_id = request.args.get('equipe_id', type=int)
    data_inicio = request.args.get('data_inicio')
    data_fim = request.args.get('data_fim')
    
    # Construir consulta
    query = db.session.query(
        db.func.date(RegistroMaterial.data_registro).label('data'),
        Equipe.id.label('equipe_id'),
        Equipe.nome.label('equipe_nome'),
        Material.id.label('material_id'),
        Material.nome.label('material_nome'),
        Material.unidade_medida,
        db.func.sum(RegistroMaterial.quantidade).label('quantidade_total')
    ).join(
        Material, RegistroMaterial.material_id == Material.id
    ).join(
        Equipe, RegistroMaterial.equipe_id == Equipe.id
    )
    
    # Aplicar filtros
    if equipe_id:
        query = query.filter(RegistroMaterial.equipe_id == equipe_id)
    if data_inicio:
        try:
            data_inicio_dt = datetime.strptime(data_inicio, '%Y-%m-%d')
            query = query.filter(RegistroMaterial.data_registro >= data_inicio_dt)
        except ValueError:
            return jsonify({'erro': 'Formato de data inválido para data_inicio. Use YYYY-MM-DD.'}), 400
    if data_fim:
        try:
            data_fim_dt = datetime.strptime(data_fim, '%Y-%m-%d')
            query = query.filter(RegistroMaterial.data_registro <= data_fim_dt)
        except ValueError:
            return jsonify({'erro': 'Formato de data inválido para data_fim. Use YYYY-MM-DD.'}), 400
    
    # Agrupar resultados
    query = query.group_by(
        db.func.date(RegistroMaterial.data_registro),
        Equipe.id, Equipe.nome,
        Material.id, Material.nome, Material.unidade_medida
    ).order_by(
        db.func.date(RegistroMaterial.data_registro),
        Equipe.nome,
        Material.nome
    )
    
    # Executar consulta
    resultados = query.all()
    
    # Formatar resultados para resposta
    if formato == 'json':
        dados = []
        for r in resultados:
            dados.append({
                'data': r.data.strftime('%Y-%m-%d'),
                'equipe_id': r.equipe_id,
                'equipe_nome': r.equipe_nome,
                'material_id': r.material_id,
                'material_nome': r.material_nome,
                'unidade_medida': r.unidade_medida,
                'quantidade_total': float(r.quantidade_total)
            })
        return jsonify(dados)
    elif formato == 'csv':
        return gerar_csv_material_por_dia(resultados)
    elif formato == 'pdf':
        return gerar_pdf_material_por_dia(resultados)
    else:
        return jsonify({'erro': 'Formato não suportado. Use json, csv ou pdf.'}), 400

def gerar_csv_material_por_dia(resultados):
    """Gera relatório em formato CSV"""
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Cabeçalho
    writer.writerow(['Data', 'Equipe', 'Material', 'Unidade', 'Quantidade Total'])
    
    # Dados
    for r in resultados:
        writer.writerow([
            r.data.strftime('%d/%m/%Y'),
            r.equipe_nome,
            r.material_nome,
            r.unidade_medida,
            r.quantidade_total
        ])
    
    output.seek(0)
    
    response = make_response(output.getvalue())
    response.headers['Content-Disposition'] = f'attachment; filename=material_por_dia_{datetime.now().strftime("%Y%m%d")}.csv'
    response.headers['Content-type'] = 'text/csv'
    
    return response

def gerar_pdf_material_por_dia(resultados):
    """Gera relatório em formato PDF"""
    pdf = FPDF()
    pdf.add_page()
    
    # Título
    pdf.set_font('Arial', 'B', 16)
    pdf.cell(0, 10, 'Relatório de Material por Dia', 0, 1, 'C')
    pdf.set_font('Arial', 'I', 10)
    pdf.cell(0, 10, f'Gerado em: {datetime.now().strftime("%d/%m/%Y %H:%M:%S")}', 0, 1, 'C')
    pdf.ln(5)
    
    # Agrupar por data e equipe para melhor visualização
    dados_agrupados = {}
    for r in resultados:
        data_str = r.data.strftime('%Y-%m-%d')
        chave = f"{data_str}_{r.equipe_id}"
        
        if chave not in dados_agrupados:
            dados_agrupados[chave] = {
                'data': r.data.strftime('%d/%m/%Y'),
                'equipe_nome': r.equipe_nome,
                'materiais': []
            }
        
        dados_agrupados[chave]['materiais'].append({
            'material_nome': r.material_nome,
            'unidade_medida': r.unidade_medida,
            'quantidade_total': r.quantidade_total
        })
    
    # Ordenar por data
    chaves_ordenadas = sorted(dados_agrupados.keys())
    
    # Dados
    pdf.set_font('Arial', 'B', 10)
    
    for chave in chaves_ordenadas:
        grupo = dados_agrupados[chave]
        
        # Cabeçalho do grupo
        pdf.set_fill_color(230, 230, 230)
        pdf.cell(30, 7, grupo['data'], 1, 0, 'C', True)
        pdf.cell(70, 7, grupo['equipe_nome'], 1, 1, 'L', True)
        
        # Cabeçalho da tabela de materiais
        pdf.set_font('Arial', 'B', 9)
        pdf.cell(10, 7, '', 0, 0)  # Espaço para indentação
        pdf.cell(70, 7, 'Material', 1, 0, 'C')
        pdf.cell(30, 7, 'Unidade', 1, 0, 'C')
        pdf.cell(30, 7, 'Quantidade', 1, 1, 'C')
        
        # Dados dos materiais
        pdf.set_font('Arial', '', 9)
        for material in grupo['materiais']:
            pdf.cell(10, 6, '', 0, 0)  # Espaço para indentação
            pdf.cell(70, 6, material['material_nome'], 1, 0, 'L')
            pdf.cell(30, 6, material['unidade_medida'], 1, 0, 'C')
            pdf.cell(30, 6, str(material['quantidade_total']), 1, 1, 'R')
        
        pdf.ln(5)  # Espaço entre grupos
    
    # Gerar PDF
    pdf_output = pdf.output(dest='S').encode('latin1')
    
    response = make_response(pdf_output)
    response.headers['Content-Disposition'] = f'attachment; filename=material_por_dia_{datetime.now().strftime("%Y%m%d")}.pdf'
    response.headers['Content-Type'] = 'application/pdf'
    
    return response

File: comentados/relatorios.js
// Funções para gerenciar relatórios

// Função para inicializar a interface de relatórios
function inicializarRelatorios() {
    // Adicionar evento ao botão de gerar relatório
    const btnGerarRelatorio = document.getElementById('btn-gerar-relatorio');
    if (btnGerarRelatorio) {
        btnGerarRelatorio.addEventListener('click', gerarRelatorioMaterialPorDia);
    }
    
    // Adicionar eventos aos botões de download
    const btnDownloadCSV = document.getElementById('btn-download-csv');
    if (btnDownloadCSV) {
        btnDownloadCSV.addEventListener('click', () => downloadRelatorio('csv'));
    }
    
    const btnDownloadPDF = document.getElementById('btn-download-pdf');
    if (btnDownloadPDF) {
        btnDownloadPDF.addEventListener('click', () => downloadRelatorio('pdf'));
    }
    
    // Carregar opções para o select de equipes
    const selectEquipe = document.getElementById('relatorio-equipe');
    if (selectEquipe) {
        carregarOpcoesSelect(selectEquipe, 'equipes');
    }
}

// Função para gerar relatório de material por dia
function gerarRelatorioMaterialPorDia() {
    const dataInicio = document.getElementById('relatorio-data-inicio').value;
    const dataFim = document.getElementById('relatorio-data-fim').value;
    const equipeId = document.getElementById('relatorio-equipe').value;
    
    // Validar datas
    if (!dataInicio || !dataFim) {
        mostrarAlerta('Selecione o período para o relatório', 'warning');
        return;
    }
    
    // Construir URL com parâmetros
    let url = '/api/relatorios/material-por-dia?';
    if (dataInicio) url += `data_inicio=${dataInicio}&`;
    if (dataFim) url += `data_fim=${dataFim}&`;
    if (equipeId) url += `equipe_id=${equipeId}&`;
    
    // Remover último &
    url = url.endsWith('&') ? url.slice(0, -1) : url;
    
    // Mostrar indicador de carregamento
    const resultadoContainer = document.getElementById('relatorio-resultado');
    if (resultadoContainer) {
        resultadoContainer.innerHTML = '<div class="loading">Carregando relatório...</div>';
    }
    
    // Fazer requisição
    fetch(url)
        .then(response => {
            if (!response.ok) {
                throw new Error('Erro ao carregar relatório');
            }
            return response.json();
        })
        .then(data => {
            renderizarRelatorioMaterialPorDia(data);
            
            // Habilitar botões de download
            document.getElementById('btn-download-csv').disabled = false;
            document.getElementById('btn-download-pdf').disabled = false;
        })
        .catch(error => {
            mostrarAlerta('Erro ao carregar relatório: ' + error.message, 'error');
            if (resultadoContainer) {
                resultadoContainer.innerHTML = '<div class="error">Erro ao carregar relatório</div>';
            }
        });
}

// Função para renderizar o relatório de material por dia
function renderizarRelatorioMaterialPorDia(dados) {
    const resultadoContainer = document.getElementById('relatorio-resultado');
    if (!resultadoContainer) return;
    
    // Se não houver dados
    if (!dados || dados.length === 0) {
        resultadoContainer.innerHTML = '<div class="info">Nenhum registro encontrado para o período selecionado</div>';
        return;
    }
    
    // Agrupar dados por data e equipe
    const dadosAgrupados = {};
    dados.forEach(item => {
        const chave = `${item.data}_${item.equipe_id}`;
        
        if (!dadosAgrupados[chave]) {
            dadosAgrupados[chave] = {
                data: item.data,
                data_formatada: formatarData(item.data),
                equipe_id: item.equipe_id,
                equipe_nome: item.equipe_nome,
                materiais: []
            };
        }
        
        dadosAgrupados[chave].materiais.push({
            material_id: item.material_id,
            material_nome: item.material_nome,
            unidade_medida: item.unidade_medida,
            quantidade_total: item.quantidade_total
        });
    });
    
    // Ordenar chaves por data
    const chavesOrdenadas = Object.keys(dadosAgrupados).sort();
    
    // Criar HTML do relatório
    let html = `
        <h3>Relatório de Material por Dia</h3>
        <p>Período: ${formatarData(dados[0].data)} a ${formatarData(dados[dados.length - 1].data)}</p>
    `;
    
    // Criar tabela para cada grupo (data + equipe)
    chavesOrdenadas.forEach(chave => {
        const grupo = dadosAgrupados[chave];
        
        html += `
            <div class="relatorio-grupo">
                <div class="relatorio-grupo-header">
                    <div class="data">${grupo.data_formatada}</div>
                    <div class="equipe">${grupo.equipe_nome}</div>
                </div>
                <table class="table">
                    <thead>
                        <tr>
                            <th>Material</th>
                            <th>Unidade</th>
                            <th>Quantidade</th>
                        </tr>
                    </thead>
                    <tbody>
        `;
        
        // Adicionar linhas para cada material
        grupo.materiais.forEach(material => {
            html += `
                <tr>
                    <td>${material.material_nome}</td>
                    <td>${material.unidade_medida}</td>
                    <td class="quantidade">${material.quantidade_total}</td>
                </tr>
            `;
        });
        
        // Calcular total para materiais com mesma unidade
        const totaisPorUnidade = {};
        grupo.materiais.forEach(material => {
            if (!totaisPorUnidade[material.unidade_medida]) {
                totaisPorUnidade[material.unidade_medida] = 0;
            }
            totaisPorUnidade[material.unidade_medida] += material.quantidade_total;
        });
        
        // Adicionar linha de total para cada unidade
        Object.keys(totaisPorUnidade).forEach(unidade => {
            html += `
                <tr class="total">
                    <td colspan="2">Total (${unidade})</td>
                    <td class="quantidade">${totaisPorUnidade[unidade]}</td>
                </tr>
            `;
        });
        
        html += `
                    </tbody>
                </table>
            </div>
        `;
    });
    
    resultadoContainer.innerHTML = html;
}

// Função para download de relatório
function downloadRelatorio(formato) {
    const dataInicio = document.getElementById('relatorio-data-inicio').value;
    const dataFim = document.getElementById('relatorio-data-fim').value;
    const equipeId = document.getElementById('relatorio-equipe').value;
    
    // Validar datas
    if (!dataInicio || !dataFim) {
        mostrarAlerta('Selecione o período para o relatório', 'warning');
        return;
    }
    
    // Construir URL com parâmetros
    let url = `/api/relatorios/material-por-dia?formato=${formato}`;
    if (dataInicio) url += `&data_inicio=${dataInicio}`;
    if (dataFim) url += `&data_fim=${dataFim}`;
    if (equipeId) url += `&equipe_id=${equipeId}`;
    
    // Abrir em nova aba para download
    window.open(url, '_blank');
}

// Função auxiliar para formatar data
function formatarData(dataStr) {
    if (!dataStr) return '';
    
    try {
        const partes = dataStr.split('-');
        return `${partes[2]}/${partes[1]}/${partes[0]}`;
    } catch (e) {
        return dataStr;
    }
}

// Adicionar ao evento DOMContentLoaded
document.addEventListener('DOMContentLoaded', function() {
    // Inicializar relatórios se estiver na página de relatórios
    if (document.getElementById('relatorios')) {
        inicializarRelatorios();
    }
});
